nixos/modules/security/pam.nix

```diff
@@ -354,7 +354,7 @@ let
             || cfg.enableGnomeKeyring
             || cfg.googleAuthenticator.enable
             || cfg.duoSecurity.enable)) ''
-              auth required pam_unix.so ${optionalString cfg.allowNullPassword "nullok"} likeauth
+              auth requisite pam_unix.so ${optionalString cfg.allowNullPassword "nullok"} likeauth
```

Comment:
Not sure about this, you should read the comment above (line 344-349). Also pinging the original author @obadz and another person that seems to know their ways around PAM, @flokli

Reply:
Yeah, I forgot to change the comment, if folks end up agreeing I will edit the comment, but can wait until they get a chance to look.

I think requisite brings the right behavior when used in conjunction with additional challenge/response modules -- the behavior before this change isn't what I would have wanted. (if password was incorrect, required modules after this line were still being executed)

Reply:
I mostly did dig into the `account` module interface, less into the `auth` module interface.
So I'm not so sure about this either, and whether this might break other authentication modules (like pam_ldap).

Why is this so complicated in first place? Why do we call all pam modules two times? I wonder whether we might align with other distros do it here instead, they don't seem to load everything twice.

About avoiding to run duologin even when the password is wrong, pam writes the following for "required":

> If a 'required' module returns a status that is not 'success',
the operation will ultimately fail, but only after the modules below
it are invoked. This seems senseless at first glance I suppose, but
it serves the purpose of always acting the same way from the point
of view of the user trying to utilize the service. The net effect is
that it becomes impossible for a potential cracker to determine
which module caused the failure - and the less information a
malicious user has about your system, the better.

Which kinda makes sense too - so my suggestion would be to stick with keeping "auth required" for the scope of this PR, then as followup, cleanup the configuration duplication and add a early bailout switch for exiting before any 2FA, but after all primary authentication happened.

Reply:
This sounds good to me.  I will revert the part of the change with ‘auth required’ and think through a follow up change to pam.

For changes that span across packages in the nixos module is there a good place for me to post a proposal for comment before writing them — or is it preferred to have code first?

Reply:
Ok, this is ready for review again, thanks!